### PR TITLE
pad for chunk

### PR DESCRIPTION
**Pull request by rllin-fathom**
<PR title>

**ASANA TASK LINK:**
https://app.asana.com/0/1137246510213018/1145054155499130
**DESIGN DOC LINK:**
https://docs.google.com/document/d/1qNCauZ8syxS9-_5I6xS-u7T8gUeT0OBFg3JV9mjbk60/edit#heading=h.84y283j1vvo3

DETAILS
-------
1. paired with https://github.com/medicode/diseaseTools/pull/4809
2. undo https://github.com/medicode/tensor2tensor/pull/138 for TPU to work and also https://github.com/medicode/tensor2tensor/pull/138 doesn't seem like it was responsible for https://app.asana.com/0/911210587841072/1142311499808660/f

CHECK LIST
----------
**TEST PLAN:**
1. ran to AUC parity for unpacked and packed

**DID I MATERIALLY UPDATE DOCKERFILE(S)?:**
N
**ANY KNOWN IMPACT TO CUSTOMER SLAs?:**
N

**RISK & IMPACT ASSESSMENT:**
minimal

### CODE REVIEW

File: tensor2tensor/data_generators/problem.py [hunk 1]
```diff
@@ -972,13 +989,22 @@ def define_shapes(example):
...
         batching_scheme = self._get_batching_scheme(hparams, num_shards)
 
+        # if GPU and we are chunking input features,
+        # we need to pad the features that we chunk to the next nearest
+        # multiple of the chunk length,
+        # this can differ on the GPU as we do have variable input lengths
         if hasattr(hparams, 'bert_max_length'):
-          dataset = dataset.map(lambda x: pad_inputs_to_chunk_len(
-            x, hparams.bert_max_length))
+          dataset = dataset.map(
```
Comment (etragas-fathom) [inline tensor2tensor/data_generators/problem.py]: Delete pad_inputs_to_chunk_len?

Comment (etragas-fathom) [reply tensor2tensor/data_generators/problem.py]: (The function)

File: tensor2tensor/data_generators/problem.py [hunk 2]
```diff
@@ -1106,12 +1135,34 @@ def pad_one_shape(shape, none_filler):
       ]
 
     for key, shape in six.iteritems(shapes_dict):
-      if key == "inputs":
-        padded_shapes[key] = pad_one_shape(shape, inputs_none_filler)
+      # TODO: rename inputs_chunk_mask as it behaves differently
```
Comment (etragas-fathom) [inline tensor2tensor/data_generators/problem.py]: Put a \# Fathom block?

File: tensor2tensor/data_generators/problem.py [hunk 3]
```diff
@@ -1340,6 +1392,9 @@ def standardize_shapes(features, batch_size=None):
       # Assert shapes are fully known
       t.get_shape().assert_is_fully_defined()
 
+  for n, t in features.items():
```
Comment (etragas-fathom) [inline tensor2tensor/data_generators/problem.py]: Hide this behind a debug flag or something similar?

File: tensor2tensor/data_generators/problem.py [hunk 4]
```diff
@@ -1088,7 +1114,10 @@ def serving_input_fn(self, hparams):
         features=features, receiver_tensors=serialized_example)
 
   def _pad_for_tpu(self, shapes_dict, hparams):
-    """Pads unknown features' dimensions for TPU."""
+    """Pads unknown features' dimensions for TPU.
+
+    NOTE: some packing specific features and shapes
```
Comment (etragas-fathom) [inline tensor2tensor/data_generators/problem.py]: Can we make a note here that packing necessarily implies chunking? (I think that is true in the below implementation)

File: tensor2tensor/data_generators/problem.py (hunk 1 above)
Comment (rllin-fathom) [reply tensor2tensor/data_generators/problem.py]: good catch

File: tensor2tensor/data_generators/problem.py (hunk 4 above)
Comment (rllin-fathom) [reply tensor2tensor/data_generators/problem.py]: yea for now packing always implies chunking (tho not the reverse)

Comment (rllin-fathom) [reply tensor2tensor/data_generators/problem.py]: added a comment

File: tensor2tensor/data_generators/problem.py [hunk 5]
```diff
@@ -1112,6 +1146,34 @@ def pad_one_shape(shape, none_filler):
         padded_shapes[key] = pad_one_shape(shape, targets_none_filler)
       else:
         padded_shapes[key] = pad_one_shape(shape, max_length)
+
+    # Fathom
+    # override shapes for packed datsets that will be chunked
+    packed_length = getattr(self, 'packed_length', False)
```
Comment (etragas-fathom) [inline tensor2tensor/data_generators/problem.py]: 👏 great override

File: tensor2tensor/bin/t2t_decoder.py [hunk 6]
```diff
@@ -65,6 +65,7 @@
 flags.DEFINE_bool("decode_interactive", False,
                   "Interactive local inference mode.")
 flags.DEFINE_integer("decode_shards", 1, "Number of decoding replicas.")
+flags.DEFINE_string("problems", "", "Problem to use in decode")
```
Comment (kevincvlam) [inline tensor2tensor/bin/t2t_decoder.py]: nit. Should this be just 'problem'?

File: tensor2tensor/data_generators/problem.py [hunk 7]
```diff
@@ -972,13 +989,22 @@ def define_shapes(example):
           dataset = dataset.padded_batch(
               batch_size, padded_shapes, drop_remainder=True)
       else:
-        # On GPU, bucket by length
+        # for unpacked datasets, bucket by length
         dataset = dataset.filter(gpu_valid_size)
         batching_scheme = self._get_batching_scheme(hparams, num_shards)
 
+        # if GPU and we are chunking input features,
```
Comment (kevincvlam) [inline tensor2tensor/data_generators/problem.py]: The comment says 'if GPU' but there isn't a check for GPU? How does that work ooc?

File: tensor2tensor/data_generators/problem.py [hunk 8]
```diff
@@ -1088,7 +1114,15 @@ def serving_input_fn(self, hparams):
         features=features, receiver_tensors=serialized_example)
 
   def _pad_for_tpu(self, shapes_dict, hparams):
-    """Pads unknown features' dimensions for TPU."""
+    """Pads unknown features' dimensions for TPU.
+
+    If self.packed_length is specified, this will handling packed datasets
```
Comment (kevincvlam) [inline tensor2tensor/data_generators/problem.py]: nit. Maybe a short explanation of what 'handling packed datasets' means?

File: tensor2tensor/data_generators/problem.py [hunk 9]
```diff
@@ -1321,15 +1383,18 @@ def _summarize_features(features, num_shards=1):
 def standardize_shapes(features, batch_size=None):
   """Set the right shapes for the features."""
 
-  for fname in ["inputs", "targets"]:
-    if fname not in features:
+  # TODO: think of better abstraction for this, presumably should
```
Comment (kevincvlam) [inline tensor2tensor/data_generators/problem.py]: nit. explanation of why only these features/not other features

File: tensor2tensor/data_generators/problem.py (hunk 7 above)
Comment (rllin-fathom) [reply tensor2tensor/data_generators/problem.py]: updated comment, but added an assert above that if TPU then check that we are packed

File: tensor2tensor/data_generators/problem.py (hunk 8 above)
Comment (rllin-fathom) [reply tensor2tensor/data_generators/problem.py]: updated

File: tensor2tensor/data_generators/problem.py (hunk 9 above)
Comment (rllin-fathom) [reply tensor2tensor/data_generators/problem.py]: done

File: tensor2tensor/layers/common_hparams.py [hunk 10]
```diff
@@ -67,6 +67,10 @@ def basic_params1():
       optimizer_adafactor_multiply_by_parameter_scale=True,
       # Number of accumulating steps for multi step optimizers.
       optimizer_multistep_accumulate_steps=None,
+      # Whether to zero gradients that were not computed, so that the
+      # appropriate slots are created. Useful for sharing checkpoints between
+      # models with different sets of heads.
+      optimizer_zero_grads=False,
```
Comment (etragas-fathom) [inline tensor2tensor/layers/common_hparams.py]: This is pretty cool, but where / when do we use it? If we never use it in prod atm can we make a note that it's experimental? E.g debug_optimizer_zero_grads

File: tensor2tensor/utils/optimize.py [hunk 11]
```diff
@@ -113,36 +113,44 @@ def __init__(self, optimizer_name, lr, hparams, use_tpu=False):  # pylint: disab
...
           We need `use_resource=False` in model_fn in utils/t2t_model.py
           and the old version of cast_grad here.
           Without both of these changes, we are very slow with
           large word embeddings on the CPU.
 
+      Sept 30 2019: We tried removing this since we are off word embeddings
+          but slowdown seems to still be around
       """
-      if v is None or g is None:
+      if v is not None or g is None:
```
Comment (etragas-fathom) [inline tensor2tensor/utils/optimize.py]: Should this not be here for sure? Just checking

File: tensor2tensor/utils/optimize.py [hunk 12]
```diff
@@ -113,36 +113,44 @@ def __init__(self, optimizer_name, lr, hparams, use_tpu=False):  # pylint: disab
...
 
+    self._zero_grads = hparams.optimizer_zero_grads
+
   def compute_gradients(self, loss, var_list=None, **kwargs):  # pylint: disable=arguments-differ
     gradients = self._opt.compute_gradients(loss, var_list, **kwargs)
-    def cast_grad(g, v):
+    def cast_grad_tpu(g, v):
+      if v is not None and g is not None:
+        g = common_layers.cast_like(g, v)
+      if self._zero_grads and g is None:
```
Comment (etragas-fathom) [inline tensor2tensor/utils/optimize.py]: elif? Or do we expect graph to prune out the cast_like above?

File: tensor2tensor/layers/common_hparams.py (hunk 10 above)
Comment (etragas-fathom) [reply tensor2tensor/layers/common_hparams.py]: Is this a Fathom feature? Is so #FATHOM around this?

File: tensor2tensor/utils/t2t_model.py [hunk 13]
```diff
@@ -325,12 +325,13 @@ def model_fn_sharded(self, sharded_features):
...
-    # large word embeddings on the CPU.
-    #with tf.variable_scope(tf.get_variable_scope(), use_resource=True) as vs:
-    with tf.variable_scope(tf.get_variable_scope(), use_resource=False) as vs:
+    # We used to need `use_resource=False` here
+    # but we don't use word embeddings on the CPU anymore
+    # NOTE: seems like there is still some slow down on GPU without word embeddings#
+    # use_resource = True when on TPU
+    # use_resource = False when on TPU
+    use_resource = common_layers.is_xla_compiled()
+    with tf.variable_scope(tf.get_variable_scope(), use_resource=use_resource) as vs:
```
Comment (etragas-fathom) [inline tensor2tensor/utils/t2t_model.py]: 👍

File: tensor2tensor/utils/t2t_model.py [hunk 14]
```diff
@@ -1459,6 +1460,11 @@ def estimator_spec_eval(self, features, logits, labels, loss, losses_dict):
         eval_metrics_fn = create_tpu_eval_metrics_fn(problem, hparams)
         # For TPU, logits dict will be passed as keyword arguments to
         # eval_metrics_fn. Here we add the labels to those arguments.
+        # NOTE: for TPU, we use pack our data, here we reshape
```
Comment (etragas-fathom) [inline tensor2tensor/utils/t2t_model.py]: rewrite on *we use pack*

File: tensor2tensor/layers/common_hparams.py (hunk 10 above)
Comment (rllin-fathom) [reply tensor2tensor/layers/common_hparams.py]: oh yea this is entirely a t2t thing pulled from upstream

File: tensor2tensor/utils/optimize.py (hunk 11 above)
Comment (rllin-fathom) [reply tensor2tensor/utils/optimize.py]: oh wow, fixed

File: tensor2tensor/utils/optimize.py (hunk 12 above)
Comment (rllin-fathom) [reply tensor2tensor/utils/optimize.py]: upstream https://github.com/tensorflow/tensor2tensor/blob/1547c25571633f828ddd74accba76d07d8d043af/tensor2tensor/utils/optimize.py#L232

i think not perfectly written, but g has to be None to go through first AND second branches anyway

**Pull request merged**